### PR TITLE
participant-integration-api: Dedicated execution context for requests.

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
Found this branch kicking around on my machine during some cleanup and decided to get it working again.

This is similar (but more extensive) to similar changes made in Canton.

My hypothesis is that by creating a dedicated execution context to handle requests (but ideally not the computation associated with them), we have more control over our runtime.

Unfortunately, I don't have a decent way to profile this. The ledger-on-memory conformance tests are about 5% faster on my machine, but this is probably within error margins.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/StandaloneApiServer.scala [hunk 1]
```diff
@@ -80,6 +82,9 @@ final class StandaloneApiServer(
       authorizer = new Authorizer(Clock.systemUTC.instant _, ledgerId, participantId)
       healthChecksWithIndexService = healthChecks + ("index" -> indexService)
       executionSequencerFactory <- new ExecutionSequencerFactoryOwner()
+      servicesExecutionContext <- ResourceOwner
+        .forExecutorService(() => Executors.newWorkStealingPool())
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/StandaloneApiServer.scala:86]: If we are to replace the direct execution context, a work-stealing thread pool is probably the best choice. Well done!

**Review by stefanobaghino-da — COMMENTED** [review]
In general it LGTM, but I would prefer having some form of benchmark run on this to inform our decision (and possibly a second opinion on the code itself).

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/services/ActiveContractsServiceAuthorization.scala [hunk 2]
```diff
@@ -31,7 +33,7 @@ private[daml] final class ActiveContractsServiceAuthorization(
       service.getActiveContracts)(request, responseObserver)
 
   override def bindService(): ServerServiceDefinition =
-    ActiveContractsServiceGrpc.bindService(this, DirectExecutionContext)
+    ActiveContractsServiceGrpc.bindService(this, executionContext)
```
[the remark this answers is not in the document]
Comment (SamirTalwar-DA) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/services/ActiveContractsServiceAuthorization.scala:36]: Nope, sorry. This makes a few related changes, and this isn't the only one. If you have an idea how to profile this in isolation we can try each change in turn.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/StandaloneApiServer.scala (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/StandaloneApiServer.scala:86]: I am a fan. :slightly_smiling_face:

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/services/ActiveContractsServiceAuthorization.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/services/ActiveContractsServiceAuthorization.scala:36]: > This makes a few related changes, and this isn't the only one.

I understand. I preferred to not put a comment on every single instance. 🙂

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/StandaloneApiServer.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/StandaloneApiServer.scala:86]: Well, what I meant is that the work-stealing pool seems well suited for the workload here, with many tasks which are likely small.

**Comment by matthiasS-da** [comment]
Reading the heading, I think we did the opposite. I observed an excessive number of threads and execution contexts for no good reason. This made thread dumps rather unwieldy. We have special tests to detect deadlocks, but the excessive number of ECs made these tests pointless. I also suspected a negative impact on performance, but was unable to verify that. 
To address these issues, I kicked out some of these execution contexts.

IMHO, a separate EC for handling requests makes sense, if the system otherwise becomes unresponsive under high load. For Canton, this is however not (yet) a problem at the moment.

My ask on the indexer/ledger API server is to allow us to inject our own execution context or, if it absolutely has to create its own EC (e.g., to access the DB), to allow us configure the number of threads from outside.

**Comment by SamirTalwar-DA** [comment]
Thanks for the clarification @matthiasS-da. I will make sure that that's an option.

**Comment by SamirTalwar-DA** [comment]
This seems to make the reset service tests even more flaky than usual. Gonna dig into that.

**Comment by SamirTalwar-DA** [comment]
Hopefully #7975 will address the failures in the reset service tests.

**Comment by oliverse-da** [comment]
In case https://github.com/digital-asset/daml/pull/7988 is merged before this one, please remember to take a look at `ApiParticipantPruningService.bindService`

**Comment by oliverse-da** [comment]
Pushed an update per https://github.com/digital-asset/daml/pull/7988#discussion_r525319097 and #7988

**Comment by SamirTalwar-DA** [comment]
Did some long-running checks on some proprietary code and we get a slight performance bump (~5-10%).

@stefanobaghino-da, mind taking another look? I can show you the numbers if you like.

**Review by tudor-da — APPROVED** [review]
Nice work

**Pull request merged**